src/MKRWAN.h

```diff
@@ -1047,79 +986,135 @@ class LoRaModem : public Stream
   {
-    data.reserve(64);
+    data.reserve(msize);
     int8_t index = -1;
     int length = 0;
+    int a = 0;
     unsigned long startMillis = millis();
     do {
       YIELD();
       while (stream.available() > 0) {
-        int a = streamRead();
+        a = stream.peek();
         if (a < 0) continue;
-        data += (char)a;
-        if (r1 && data.endsWith(r1)) {
-          index = 1;
-          goto finish;
-        } else if (r2 && data.endsWith(r2)) {
-          index = 2;
-          goto finish;
-        } else if (r3 && data.endsWith(r3)) {
-          index = 3;
-          goto finish;
-        } else if (r4 && data.endsWith(r4)) {
-          index = 4;
-          goto finish;
-        } else if (r5 && data.endsWith(r5)) {
-          index = 5;
-          goto finish;
-        } else if (r6 && data.endsWith(r6)) {
-          index = 6;
-          goto finish;
-        } else if (r7 && data.endsWith(r7)) {
-          index = 7;
-          goto finish;
-        } else if (r8 && data.endsWith(r8)) {
-          index = 8;
-          goto finish;
-        } else if (data.endsWith("+RECV=")) {
-          data = "";
-          stream.readStringUntil(',').toInt();
-          length = stream.readStringUntil('\r').toInt();
-          streamSkipUntil('\n');
-          streamSkipUntil('\n');
-          for (int i = 0; i < length;) {
-            if (stream.available()) {
-                rx.put(stream.read());
-                i++;
-            }
-          }
+        if (a == '=' || a == '\r'
+        		|| (a == '+' && data.length() > 0)) {
+			DBG("### Data string:", data);
+			if (r1 && data.endsWith(r1)) {
+			  index = 1;
+			  goto finish;
+			} else if (r2 && data.endsWith(r2)) {
+			  index = 2;
+			  goto finish;
+			} else if (r3 && data.endsWith(r3)) {
+			  index = 3;
+			  goto finish;
+			} else if (r4 && data.endsWith(r4)) {
+			  index = 4;
+			  goto finish;
+			} else if (r5 && data.endsWith(r5)) {
+			  index = 5;
+			  goto finish;
+			} else if (r6 && data.endsWith(r6)) {
+			  index = 6;
+			  goto finish;
+			} else if (r7 && data.endsWith(r7)) {
+			  index = 7;
+			  goto finish;
+			} else if (r8 && data.endsWith(r8)) {
+			  index = 8;
+			  goto finish;
+			} else if ((data.endsWith(AT_RECV)
+					|| data.endsWith(AT_RECVB)) && a == '=') {
+			  (void)stream.readStringUntil(',').toInt();
+			  length = stream.readStringUntil('\r').toInt();
+			  (void)streamSkipUntil('\n');
+			  (void)streamSkipUntil('\n');
+			  if ((uint16_t)length >= msize){
+				  DBG("### Data string too long:", data);
+				  data = "";
+				  length = 0;
+				  continue;
+			  }
+			  if (data.endsWith(AT_RECVB)){ // Binary receive
+				  char Hi = 0;
+				  for (int i = 0; i < length*2;) {
+					if (stream.available()) {
+						if (!(i%2))
+							Hi=char2int(stream.read()) * 0x10;
+						else
+							rx.put(char2int(stream.read()) + Hi);
+						i++;
+					}
+				  }
+			  }
+			  else	// String receive
```

Comment:
return value should be unique?

Reply:
Same as above..
Don't know how we might implement this in an other way.

Reply:
I checked the history and at some point, the 99 return value has been removed. I think this was because an independent receive over RX2 initiated by the server could cause an unknown send status. Ex, duty cycle exceeded RX2 works, but send does not. However, with 99 as return status, you would never know.
If a data receive was successful can also be tested by `.available()`; same for ack, can be tested by `getMsgConfirmed()`. Both do not trigger another TX/RX.

I think we can close this.